internal/client/util.go

```diff
@@ -58,3 +63,106 @@ func setDefaultHeaders(request *http.Request, apiKey string) {
 	request.Header.Set("Content-Type", "application/json")
 	request.Header.Set("Accept", "application/json")
 }
+
+// requestConfig is a configuration object for an HTTP request.
+type requestConfig struct {
+	method string
+	url    string
+	body   any
+
+	successCodes []int
+
+	apiKey string
+}
+
+var (
+	// successCodesStatusOK is a convenience variable to use for the most common
+	// success criteria.
+	successCodesStatusOK = []int{http.StatusOK}
+
+	// successCodesStatusCreated is a convenience variable to use for a common
+	// success criteria of StatusCreated.
+	successCodesStatusCreated = []int{http.StatusCreated}
+
+	// successCodesStatusNoContent is a convenience variable to use for a common
+	// success criteria of StatusNoContent.
+	successCodesStatusNoContent = []int{http.StatusNoContent}
+
+	// successCodesStatusOKOrNoContent is a convenience variable to use for a common
+	// success criteria of either StatusOK or StatusNoContent.
+	successCodesStatusOKOrNoContent = []int{http.StatusOK, http.StatusNoContent}
+)
+
+// request performs an HTTP request with the provided configuration.
+// It returns the response, or an error if the request fails.
+// The caller is responsible for closing the response body.
+func request(ctx context.Context, client *http.Client, cfg requestConfig) (*http.Response, error) {
+	var body io.Reader
+
+	if cfg.body != nil && cfg.body != http.NoBody {
+		var buf bytes.Buffer
+		if err := json.NewEncoder(&buf).Encode(cfg.body); err != nil {
+			return nil, fmt.Errorf("failed to encode body data: %w", err)
+		}
+
+		body = &buf
+	} else {
+		body = http.NoBody
+	}
+
+	req, err := http.NewRequestWithContext(ctx, cfg.method, cfg.url, body)
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %w", err)
+	}
+
+	setDefaultHeaders(req, cfg.apiKey)
+
+	// Body will be closed by the caller.
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("http error: %w", err)
+	}
+
+	success := false
+	for _, successCode := range cfg.successCodes {
+		if resp.StatusCode == successCode {
+			success = true
+
+			break
+		}
+	}
+
+	if !success {
+		return nil, fmt.Errorf("status code %s", resp.Status)
+	}
+
+	return resp, nil
+}
+
+// decodeResponseBody decodes the response body into the target object.
+// It returns an error if the decoding fails.
+func decodeResponseBody(respBody io.ReadCloser, target any) error {
+	if err := json.NewDecoder(respBody).Decode(target); err != nil {
+		return fmt.Errorf("failed to decode response: %w", err)
+	}
+
+	return nil
+}
+
+// requestWithDecodeResponse performs an HTTP request with the provided configuration,
+// and decodes the response body into the target object.
+// It returns an error if the request fails or the decoding fails.
+func requestWithDecodeResponse(ctx context.Context, client *http.Client, cfg requestConfig, target any) error {
+	resp, err := request(ctx, client, cfg)
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
```

Comment:
guess the only difference is that this `resp.Body.Close()` happens _**slightly**_ sooner, since it gets deferred from within the helper context + not the actual API method.

Reply:
@parkedwards  That's true - I considered making the defer func a return value so the caller can call it when needed, but I started with the simpler approach and there doesn't seem to be any noticeable impact with the acceptance tests. Think it's worth doing though? I'm happy either way

Reply:
im good to roll with it! i just remember going down this path a while ago, but the `defer` part was throwing me so i didn't end up touching the logic.  but it might not make that big of a difference